File: android/feature/profile/src/main/kotlin/hata/feature/profile/ui/ProfileScreen.kt
package hata.feature.profile.ui

import androidx.compose.foundation.background
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.ColumnScope
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.foundation.verticalScroll
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.automirrored.filled.ArrowBack
import androidx.compose.material.icons.filled.Person
import androidx.compose.material3.AlertDialog
import androidx.compose.material3.Button
import androidx.compose.material3.ButtonDefaults
import androidx.compose.material3.Card
import androidx.compose.material3.CardDefaults
import androidx.compose.material3.ExperimentalMaterial3Api
import androidx.compose.material3.Icon
import androidx.compose.material3.IconButton
import androidx.compose.material3.Scaffold
import androidx.compose.material3.Text
import androidx.compose.material3.TextButton
import androidx.compose.material3.TopAppBar
import androidx.compose.material3.TopAppBarDefaults
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.getValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.tooling.preview.Preview
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import androidx.lifecycle.compose.collectAsStateWithLifecycle
import androidx.lifecycle.viewmodel.compose.viewModel
import hata.ui.theme.HataAccentColors
import hata.ui.theme.HataColors
import hata.ui.utils.preview
import java.time.Instant
import java.time.ZoneId
import java.time.format.DateTimeFormatter


@Composable
fun ProfileScreen(
    vm: ProfileViewModel = viewModel(),
) {
    val state by vm.uiState.collectAsStateWithLifecycle()

    LaunchedEffect(Unit) {
        vm.onDispatch(ProfileUiAction.Init)
    }

    ProfileScreenUI(
        state = state,
        dispatch = vm::onDispatch,
    )
}

@Composable
private fun ProfileScreenUI(
    state: ProfileUiState,
    dispatch: (ProfileUiAction) -> Unit = {},
) {
    Scaffold(
        topBar = {
            TopBar(dispatch)
        },
        containerColor = HataColors.background,
    ) { paddingValues ->
        Box(
            modifier = Modifier
                .fillMaxSize()
                .padding(paddingValues),
        ) {
            when (state) {
                is ProfileUiState.Init -> {}

                is ProfileUiState.Loaded -> {
                    ProfileContent(
                        userName = state.userName,
                        serverUrl = state.serverUrl,
                        accountStatus = state.accountStatus,
                        appVersion = state.appVersion,
                        appPackage = state.appPackage,
                        sessionCreatedAt = state.sessionCreatedAt,
                        showLogoutDialog = state.showLogoutDialog,
                        onLogout = { dispatch(ProfileUiAction.Logout) },
                        onShowLogoutDialog = { dispatch(ProfileUiAction.ShowLogoutDialog) },
                        onDismissLogoutDialog = { dispatch(ProfileUiAction.DismissLogoutDialog) },
                    )
                }
            }
        }
    }
}

@Composable
@OptIn(ExperimentalMaterial3Api::class)
private fun TopBar(dispatch: (ProfileUiAction) -> Unit) {
    TopAppBar(
        title = { Text("Profile") },
        navigationIcon = {
            IconButton(onClick = { dispatch(ProfileUiAction.Back) }) {
                Icon(
                    imageVector = Icons.AutoMirrored.Filled.ArrowBack,
                    contentDescription = "Back",
                )
            }
        },
        colors = TopAppBarDefaults.topAppBarColors(
            containerColor = HataColors.background,
            titleContentColor = Color.White,
            navigationIconContentColor = Color.White,
        ),
    )
}

@Composable
private fun ProfileContent(
    userName: String,
    serverUrl: String,
    accountStatus: String,
    appVersion: String,
    appPackage: String,
    sessionCreatedAt: Long,
    showLogoutDialog: Boolean,
    onLogout: () -> Unit,
    onShowLogoutDialog: () -> Unit,
    onDismissLogoutDialog: () -> Unit,
) {
    Box(
        modifier = Modifier.fillMaxSize(),
    ) {
        Column(
            modifier = Modifier
                .fillMaxSize()
                .verticalScroll(rememberScrollState())
                .padding(start = 32.dp, end = 32.dp, top = 0.dp, bottom = 120.dp),
            horizontalAlignment = Alignment.CenterHorizontally,
        ) {
            Spacer(modifier = Modifier.height(16.dp))

            Column(
                horizontalAlignment = Alignment.CenterHorizontally,
            ) {
                UserAvatar()

                Spacer(modifier = Modifier.height(24.dp))

                UserNameDisplay(userName = userName)
            }

            Spacer(modifier = Modifier.height(48.dp))

            ProfileInfoSections(
                serverUrl = serverUrl,
                accountStatus = accountStatus,
                appVersion = appVersion,
                appPackage = appPackage,
                sessionCreatedAt = sessionCreatedAt,
            )
        }

        LogoutButton(
            onClick = onShowLogoutDialog,
            modifier = Modifier
                .align(Alignment.BottomCenter)
                .padding(start = 32.dp, end = 32.dp, bottom = 24.dp),
        )
    }

    // Logout Confirmation Dialog
    if (showLogoutDialog) {
        LogoutDialog(
            onConfirm = {
                onDismissLogoutDialog()
                onLogout()
            },
            onDismiss = onDismissLogoutDialog,
        )
    }
}

@Composable
private fun UserAvatar() {
    Box(
        modifier = Modifier
            .size(120.dp)
            .clip(CircleShape)
            .background(HataAccentColors.avatarBackground),
        contentAlignment = Alignment.Center,
    ) {
        Icon(
            imageVector = Icons.Default.Person,
            contentDescription = "Profile",
            tint = HataAccentColors.avatarIcon,
            modifier = Modifier.size(64.dp),
        )
    }
}

@Composable
private fun UserNameDisplay(userName: String) {
    Text(
        text = userName,
        color = Color.White,
        fontSize = 32.sp,
        fontWeight = FontWeight.Bold,
    )
}

@Composable
private fun ProfileInfoSections(
    serverUrl: String,
    accountStatus: String,
    appVersion: String,
    appPackage: String,
    sessionCreatedAt: Long,
) {
    Column(
        modifier = Modifier.fillMaxWidth(),
    ) {
        SectionTitle(title = "User")
        InfoCard {
            InfoRow(label = "Created", value = formatSessionCreatedAt(sessionCreatedAt))
        }

        Spacer(modifier = Modifier.height(20.dp))

        SectionTitle(title = "Server")
        InfoCard {
            InfoRow(label = "URL", value = serverUrl)
            Spacer(modifier = Modifier.height(12.dp))
            StatusRow(label = "Status", status = accountStatus)
        }

        Spacer(modifier = Modifier.height(20.dp))

        SectionTitle(title = "App")
        InfoCard {
            InfoRow(label = "Package", value = appPackage)
            Spacer(modifier = Modifier.height(12.dp))
            InfoRow(label = "Version", value = appVersion)
        }
    }
}

@Composable
private fun InfoCard(content: @Composable ColumnScope.() -> Unit) {
    Card(
        modifier = Modifier.fillMaxWidth(),
        colors = CardDefaults.cardColors(
            containerColor = HataColors.surface,
        ),
        shape = RoundedCornerShape(16.dp),
    ) {
        Column(
            modifier = Modifier.padding(24.dp),
            content = content,
        )
    }
}

@Composable
private fun SectionTitle(title: String) {
    Text(
        text = title,
        color = HataColors.onSurfaceVariant,
        fontSize = 12.sp,
        fontWeight = FontWeight.SemiBold,
    )
    Spacer(modifier = Modifier.height(8.dp))
}

@Composable
private fun InfoRow(label: String, value: String) {
    Row(
        modifier = Modifier.fillMaxWidth(),
        horizontalArrangement = Arrangement.SpaceBetween,
    ) {
        Text(
            text = label,
            color = HataColors.onSurfaceVariant,
            fontSize = 16.sp,
        )
        Text(
            text = value,
            color = Color.White,
            fontSize = 16.sp,
            fontWeight = FontWeight.Medium,
        )
    }
}

@Composable
private fun StatusRow(label: String, status: String) {
    Row(
        modifier = Modifier.fillMaxWidth(),
        horizontalArrangement = Arrangement.SpaceBetween,
        verticalAlignment = Alignment.CenterVertically,
    ) {
        Text(
            text = label,
            color = HataColors.onSurfaceVariant,
            fontSize = 16.sp,
        )
        Row(
            verticalAlignment = Alignment.CenterVertically,
        ) {
            StatusDot(color = statusColor(status))
            Spacer(modifier = Modifier.width(8.dp))
            Text(
                text = status,
                color = Color.White,
                fontSize = 16.sp,
                fontWeight = FontWeight.Medium,
            )
        }
    }
}

@Composable
private fun StatusDot(color: Color) {
    Box(
        modifier = Modifier
            .size(8.dp)
            .clip(CircleShape)
            .background(color),
    )
}

private fun formatSessionCreatedAt(timestamp: Long): String {
    val formatter = DateTimeFormatter.ofPattern("MMM d, yyyy • HH:mm")
    val instant = Instant.ofEpochMilli(timestamp)
    return formatter.format(instant.atZone(ZoneId.systemDefault()))
}

private fun statusColor(status: String): Color {
    return when (status.lowercase()) {
        "active" -> HataColors.primary
        "inactive" -> HataColors.error
        "warning", "pending" -> Color(0xFFF2C94C)
        else -> Color(0xFFF2C94C)
    }
}

@Composable
private fun LogoutButton(
    onClick: () -> Unit,
    modifier: Modifier = Modifier,
) {
    Button(
        onClick = onClick,
        modifier = modifier
            .fillMaxWidth()
            .height(56.dp),
        colors = ButtonDefaults.buttonColors(
            containerColor = HataColors.error,
            contentColor = HataColors.onPrimary,
        ),
        shape = RoundedCornerShape(12.dp),
    ) {
        Text(
            text = "Logout",
            fontSize = 18.sp,
            fontWeight = FontWeight.SemiBold,
        )
    }
}

@Composable
private fun LogoutDialog(
    onConfirm: () -> Unit,
    onDismiss: () -> Unit,
) {
    AlertDialog(
        onDismissRequest = onDismiss,
        title = {
            Text(
                text = "Logout",
                color = Color.White,
            )
        },
        text = {
            Text(
                text = "Are you sure you want to logout?",
                color = HataColors.onSurfaceVariant,
            )
        },
        confirmButton = {
            TextButton(
                onClick = onConfirm,
                colors = ButtonDefaults.textButtonColors(
                    contentColor = HataColors.error,
                ),
            ) {
                Text("Logout")
            }
        },
        dismissButton = {
            TextButton(
                onClick = onDismiss,
                colors = ButtonDefaults.textButtonColors(
                    contentColor = Color.White,
                ),
            ) {
                Text("Cancel")
            }
        },
        containerColor = HataColors.surface,
    )
}

@Preview
@Composable
private fun Preview_ProfileScreen() = preview {
    ProfileScreenUI(
        state = ProfileUiState.Loaded(
            userName = "Dan",
            serverUrl = "https://hata.local",
            accountStatus = "Active",
            appVersion = "1.0.0",
            appPackage = "space.dector.hata",
            sessionCreatedAt = System.currentTimeMillis(),
            showLogoutDialog = false,
        ),
    )
}

@Preview
@Composable
private fun Preview_UserAvatar() = preview {
    UserAvatar()
}

@Preview
@Composable
private fun Preview_UserNameDisplay() = preview {
    UserNameDisplay(userName = "Dan")
}

@Preview
@Composable
private fun Preview_ProfileInfoSections() = preview {
    ProfileInfoSections(
        serverUrl = "https://hata.local",
        accountStatus = "Active",
        appVersion = "1.0.0",
        appPackage = "space.dector.hata",
        sessionCreatedAt = System.currentTimeMillis(),
    )
}

@Preview
@Composable
private fun Preview_LogoutButton() = preview {
    LogoutButton(onClick = {})
}

@Preview
@Composable
private fun Preview_LogoutDialog() = preview(pad = 4.dp) {
    LogoutDialog(
        onConfirm = {},
        onDismiss = {},
    )
}

File: app/lib/screens/profile_screen.dart
import 'package:flutter/material.dart';
import 'package:package_info_plus/package_info_plus.dart';

import '../api/api.dart';
import '../models/server_info.dart';
import '../session/session.dart';
import '../theme/hata_colors.dart';

class ProfileScreen extends StatefulWidget {
  const ProfileScreen({
    super.key,
    required this.session,
    required this.apiClient,
    required this.onLogout,
  });

  final Session session;
  final Api apiClient;
  final Future<void> Function() onLogout;

  @override
  State<ProfileScreen> createState() => _ProfileScreenState();
}

class _ProfileScreenState extends State<ProfileScreen> {
  PackageInfo? _packageInfo;
  ServerInfo? _serverInfo;

  @override
  void initState() {
    super.initState();
    _loadPackageInfo();
    _loadServerInfo();
  }

  Future<void> _loadPackageInfo() async {
    try {
      final packageInfo = await PackageInfo.fromPlatform();
      if (!mounted) return;
      setState(() => _packageInfo = packageInfo);
    } catch (_) {
      // Keep fallback values visible if platform package info is unavailable.
    }
  }

  Future<void> _loadServerInfo() async {
    try {
      final serverInfo = await widget.apiClient.ping(widget.session.serverUrl);
      if (!mounted) return;
      setState(() => _serverInfo = serverInfo);
    } catch (_) {
      // Keep fallback values visible if the server cannot be reached.
    }
  }

  Future<void> _showLogoutDialog() async {
    final confirmed = await showDialog<bool>(
      context: context,
      builder: (context) => const _LogoutDialog(),
    );
    if (confirmed != true || !mounted) return;

    await widget.onLogout();
    if (!mounted) return;
    Navigator.of(context).popUntil((route) => route.isFirst);
  }

  @override
  Widget build(BuildContext context) {
    final session = widget.session;
    final userName = (session.displayName?.trim().isNotEmpty ?? false)
        ? session.displayName!.trim()
        : session.username;
    final packageInfo = _packageInfo;

    return Scaffold(
      backgroundColor: HataColors.background,
      appBar: AppBar(
        backgroundColor: HataColors.background,
        foregroundColor: Colors.white,
        title: const Text('Profile'),
      ),
      body: SafeArea(
        child: Stack(
          children: [
            SingleChildScrollView(
              padding: const EdgeInsets.fromLTRB(32, 16, 32, 120),
              child: Column(
                crossAxisAlignment: CrossAxisAlignment.center,
                children: [
                  const _UserAvatar(),
                  const SizedBox(height: 24),
                  Text(
                    userName,
                    textAlign: TextAlign.center,
                    style: const TextStyle(
                      color: Colors.white,
                      fontSize: 32,
                      fontWeight: FontWeight.w700,
                    ),
                  ),
                  const SizedBox(height: 48),
                  _ProfileInfoSections(
                    serverUrl: session.serverUrl,
                    serverVersion: _serverInfo?.version ?? '—',
                    accountStatus: session.isExpired ? 'Inactive' : 'Active',
                    appPackage: packageInfo?.packageName ?? '—',
                    appVersion: packageInfo?.version ?? '—',
                    sessionCreatedAt: session.createdAt,
                  ),
                ],
              ),
            ),
            Align(
              alignment: Alignment.bottomCenter,
              child: Padding(
                padding: const EdgeInsets.fromLTRB(32, 0, 32, 24),
                child: _LogoutButton(onPressed: _showLogoutDialog),
              ),
            ),
          ],
        ),
      ),
    );
  }
}

class _UserAvatar extends StatelessWidget {
  const _UserAvatar();

  @override
  Widget build(BuildContext context) {
    return const CircleAvatar(
      radius: 60,
      backgroundColor: HataColors.avatarBackground,
      child: Icon(Icons.person, color: HataColors.avatarIcon, size: 64),
    );
  }
}

class _ProfileInfoSections extends StatelessWidget {
  const _ProfileInfoSections({
    required this.serverUrl,
    required this.serverVersion,
    required this.accountStatus,
    required this.appVersion,
    required this.appPackage,
    required this.sessionCreatedAt,
  });

  final String serverUrl;
  final String serverVersion;
  final String accountStatus;
  final String appVersion;
  final String appPackage;
  final DateTime sessionCreatedAt;

  @override
  Widget build(BuildContext context) {
    return Column(
      crossAxisAlignment: CrossAxisAlignment.stretch,
      children: [
        const _SectionTitle('User'),
        _InfoCard(
          child: _InfoRow(
            label: 'Created',
            value: _formatSessionCreatedAt(sessionCreatedAt),
          ),
        ),
        const SizedBox(height: 20),
        const _SectionTitle('Server'),
        _InfoCard(
          child: Column(
            children: [
              _InfoRow(label: 'URL', value: serverUrl),
              const SizedBox(height: 12),
              _InfoRow(label: 'Version', value: serverVersion),
              const SizedBox(height: 12),
              _StatusRow(label: 'Status', status: accountStatus),
            ],
          ),
        ),
        const SizedBox(height: 20),
        const _SectionTitle('App'),
        _InfoCard(
          child: Column(
            children: [
              _InfoRow(label: 'Package', value: appPackage),
              const SizedBox(height: 12),
              _InfoRow(label: 'Version', value: appVersion),
            ],
          ),
        ),
      ],
    );
  }
}

class _SectionTitle extends StatelessWidget {
  const _SectionTitle(this.title);

  final String title;

  @override
  Widget build(BuildContext context) {
    return Padding(
      padding: const EdgeInsets.only(bottom: 8),
      child: Text(
        title,
        style: const TextStyle(
          color: HataColors.onSurfaceVariant,
          fontSize: 12,
          fontWeight: FontWeight.w600,
        ),
      ),
    );
  }
}

class _InfoCard extends StatelessWidget {
  const _InfoCard({required this.child});

  final Widget child;

  @override
  Widget build(BuildContext context) {
    return Card(
      margin: EdgeInsets.zero,
      color: HataColors.surface,
      shape: RoundedRectangleBorder(borderRadius: BorderRadius.circular(16)),
      child: Padding(padding: const EdgeInsets.all(24), child: child),
    );
  }
}

class _InfoRow extends StatelessWidget {
  const _InfoRow({required this.label, required this.value});

  final String label;
  final String value;

  @override
  Widget build(BuildContext context) {
    return Row(
      crossAxisAlignment: CrossAxisAlignment.start,
      children: [
        Text(
          label,
          style: const TextStyle(
            color: HataColors.onSurfaceVariant,
            fontSize: 16,
          ),
        ),
        const SizedBox(width: 16),
        Expanded(
          child: Text(
            value,
            textAlign: TextAlign.end,
            style: const TextStyle(
              color: Colors.white,
              fontSize: 16,
              fontWeight: FontWeight.w500,
            ),
          ),
        ),
      ],
    );
  }
}

class _StatusRow extends StatelessWidget {
  const _StatusRow({required this.label, required this.status});

  final String label;
  final String status;

  @override
  Widget build(BuildContext context) {
    return Row(
      crossAxisAlignment: CrossAxisAlignment.center,
      children: [
        Text(
          label,
          style: const TextStyle(
            color: HataColors.onSurfaceVariant,
            fontSize: 16,
          ),
        ),
        const Spacer(),
        Container(
          width: 8,
          height: 8,
          decoration: BoxDecoration(
            color: _statusColor(status),
            shape: BoxShape.circle,
          ),
        ),
        const SizedBox(width: 8),
        Text(
          status,
          style: const TextStyle(
            color: Colors.white,
            fontSize: 16,
            fontWeight: FontWeight.w500,
          ),
        ),
      ],
    );
  }
}

class _LogoutButton extends StatelessWidget {
  const _LogoutButton({required this.onPressed});

  final VoidCallback onPressed;

  @override
  Widget build(BuildContext context) {
    return SizedBox(
      width: double.infinity,
      height: 56,
      child: FilledButton(
        onPressed: onPressed,
        style: FilledButton.styleFrom(
          backgroundColor: HataColors.error,
          foregroundColor: HataColors.onPrimary,
          shape: RoundedRectangleBorder(
            borderRadius: BorderRadius.circular(12),
          ),
        ),
        child: const Text(
          'Logout',
          style: TextStyle(fontSize: 18, fontWeight: FontWeight.w600),
        ),
      ),
    );
  }
}

class _LogoutDialog extends StatelessWidget {
  const _LogoutDialog();

  @override
  Widget build(BuildContext context) {
    return AlertDialog(
      backgroundColor: HataColors.surface,
      title: const Text('Logout', style: TextStyle(color: Colors.white)),
      content: const Text(
        'Are you sure you want to logout?',
        style: TextStyle(color: HataColors.onSurfaceVariant),
      ),
      actions: [
        TextButton(
          onPressed: () => Navigator.of(context).pop(true),
          child: const Text(
            'Logout',
            style: TextStyle(color: HataColors.error),
          ),
        ),
        TextButton(
          onPressed: () => Navigator.of(context).pop(false),
          child: const Text('Cancel', style: TextStyle(color: Colors.white)),
        ),
      ],
    );
  }
}

String _formatSessionCreatedAt(DateTime value) {
  final local = value.toLocal();
  final month = const [
    'Jan',
    'Feb',
    'Mar',
    'Apr',
    'May',
    'Jun',
    'Jul',
    'Aug',
    'Sep',
    'Oct',
    'Nov',
    'Dec',
  ][local.month - 1];
  final hour = local.hour.toString().padLeft(2, '0');
  final minute = local.minute.toString().padLeft(2, '0');
  return '$month ${local.day}, ${local.year} • $hour:$minute';
}

Color _statusColor(String status) {
  switch (status.toLowerCase()) {
    case 'active':
      return HataColors.primary;
    case 'inactive':
      return HataColors.error;
    case 'warning':
    case 'pending':
    default:
      return const Color(0xFFF2C94C);
  }
}
